test_plans/spec-constants.md

```diff
@@ -0,0 +1,540 @@
+
+* In one translation unit:
+  - Define a `specialization_id` variable in the global namespace for the
+    tested type.  The variable's default value has some non-zero value.
+* In a second translation unit:
+  - Declare an external reference to the `specialization_id` variable.
+  - Create a `queue` from the tested device and call `queue::submit()`.
+  - Set the value of the spec constant via
+    `handler::set_specialization_constant()`.
+  - Submit a kernel via `handler::single_task()`.
+  - Read the value of the spec constant via
+    `kernel_handler::get_specialization_constant()` and make sure we get the
+    same value back.
+
+### Spec constant defined in another translation unit and set via kernel_bundle
+
+* Same test as above, except set spec constants in a kernel_bundle,
+  build the bundle and register the bundle with a handler.
+
+### Basic tests with kernel_bundle for all kernel bundle states
+
+All of the following basic tests have these initial steps:
+
+* Declare a `specialization_id` variable in the global namespace for the
+  tested type.  The variable's default value has some non-zero value.
+* Create a `queue` from the tested device and call `queue::submit()`.
+* Create `kernel_bundle` using all kernel bundle states
+  `State` == (`bundle_state::input`, `bundle_state::object`, `bundle_state::executable`)
+
+#### Read a spec constant from a kernel_bundle without writing its value
+
+* Call `kernel_bundle::get_specialization_constant()` and make sure we get the
+  default value.
+* No kernel is submitted.
+
+#### Read a spec constant from a joined kernel_bundle without writing its value
+
+* Join bundle with another bundle.
+* Call `kernel_bundle::get_specialization_constant()` and make sure we get the
+  default value.
+* No kernel is submitted.
+
+### Tests for with kernel_bundle specific to bundle_state::input
+
+All of the following basic tests have these initial steps:
+
+* Declare a `specialization_id` variable in the global namespace for the
+  tested type.  The variable's default value has some non-zero value.
+* Create a `queue` from the tested device and call `queue::submit()`.
+* Create `kernel_bundle` with`State` == `bundle_state::input`.
+
+#### Set the value in a kernel_bundle and then read it from the same bundle
+
+* Set the value of the spec constant via
+  `kernel_bundle::set_specialization_constant()`.
+* Call `kernel_bundle::get_specialization_constant()` and make sure we get the same
+  value back.
+* No kernel is submitted.
+
+#### Set the value in a kernel_bundle twice and then read it from the same bundle
+
+* Set the value of the spec constant via
+  `kernel_bundle::set_specialization_constant()`.
+* Set the value again to a different value.
+* Call `kernel_bundle::get_specialization_constant()` and make sure we get the second
+   value back.
+* No kernel is submitted.
+
+#### Read a spec constant from a compiled kernel_bundle without writing its value
+
+* Compile kernel_bundle.
```

Comment:
All these lines are unclear.
Do you mean just "Compile the bundle." or "Compile the bundle with `kernel_bundle::compile()`."?

Reply:
Yes, this was unclear.  I think 278702551abe5eb04b9cc53db60df01b63232693 addresses this.